File: swi_prolog/euler20.pl
/*

  Euler problem 20 in SWI Prolog

  """
  n! means n (n 1) ... 3 2 1
  
  Find the sum of the digits in the number 100!"
  """

  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my SWI Prolog page: http://www.hakank.org/swi_prolog/

*/

:- use_module(library(clpfd)).
:- use_module(hakank_utils).
:- use_module(euler_utils).

go :- 
        L = [
             euler20a %%,
             %% euler20b,
             %% euler20c,
             %% euler20d
            ],
        run_problems(L).

%%
%% 0.000s
%%
euler20a :-
        factorial2(100,F),
        digits_sum(F,Sum),        
        writeln(Sum).

%%
%% 0.000s
%%
euler20b :-
        factorial3(100,F),
        digits_sum(F,Sum),        
        writeln(Sum).

%%
%% 0.00s
%%
euler20c :-
        factorial4(100,F),
        digits_sum(F,Sum),        
        writeln(Sum).


%%
%% 0.002s
%%
euler20d :-
        n_factorial(100,F), % clpfd version
        digits_sum(F,Sum),
        writeln(Sum).

File: javascript/euler20.js
/* 

  Euler #20 in JavaScript.

  Problem 20:
  """
  n! means n (n 1) ... 3 2 1

  Find the sum of the digits in the number 100!")
  """

  This JavaScript program was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my JavaScript page: http://www.hakank.org/javascript_progs/

*/

'use strict';
const {prodN,sum2,range2,timing2} = require('./js_utils.js');


const factorialN = function(n) {
    return prodN(range2(1,n));
}

// 0ms
const euler20a = function() {
    return factorialN(100).toString().split("").map(i=>parseInt(i)).sum2();
}

timing2(euler20a);
